### PR TITLE
A preloading page

### LINKED ISSUE #3734: a preloading page before every page loads
We are preparing to participate in Google Code-in, and have reserved this issue for participants in GCI - but we'd love to have your help with another one! Please check out https://code.publiclab.org to see more

### Problem
Building a preloading page to enhance the behaviour of loading of html pages such that whenever a page loads a preloading page shows and on complete loading main html body shows.

## Solution
You may find solution here
http://www.marketingformavens.com/blog/create-a-page-loading-animation-with-ruby-on-rails-and-ajax

or here

https://stackoverflow.com/questions/31505506/how-to-show-page-loading-between-pages 
## Code here
https://github.com/publiclab/plots2/blob/cc2df02a63b72cca7b1129aa3044933abf614a50/app/views/layouts/application.html.erb#L57

## Thank you!

 

To learn how to write really great issues, which increases the chances they'll be resolved, see:

https://publiclab.org/wiki/developers#Contributing+for+non-coders

### PR DESCRIPTION
**Pull request by oorjitchowdhary**
Fixes #3734 
I'm a GCI-candidate.
Opened this pull request as #4031 had some problems...
@jywarren hope this works.. 😄 
Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!

### CODE REVIEW

**Comment by oorjitchowdhary** [comment]
@jywarren Pls check it once..

**Comment by oorjitchowdhary** [comment]
@publiclab/reviewers I think this can be merged?

**Comment by SidharthBansal** [comment]
@oorjitchowdhary can you please explain me the problem. 
The problem is not clear to me

**Comment by jywarren** [comment]
Hi, there is more context and discussion in https://github.com/publiclab/plots2/pull/4031, but please be aware you can force push commits to the same branch and then the discussion in the PR is not lost. But thank you!

I would like to try this out on unstable before merging it but do not have any more time today, sadly -- i want to confirm that it can never get into a "stuck" state due to poor connectivity, where the page content that has loaded is covered by the spinner. 

Thanks!

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -78,10 +78,40 @@
   <link href='https://api.mapbox.com/mapbox.js/plugins/leaflet-fullscreen/v1.0.1/leaflet.fullscreen.css' rel='stylesheet' />
   <script src="https://cdn.jsdelivr.net/npm/leaflet-search@2.4.0/dist/leaflet-search.min.js"></script>
   <link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/npm/leaflet-search@2.4.0/dist/leaflet-search.min.css">
+<link rel="stylesheet" type="text/css" href="app/assets/stylesheets/application.css">
   </head>
 
-  <body>
-
+<div id="loader-wrapper">
```
Comment (jywarren) [inline app/views/layouts/application.html.erb]: Also, i'd like to suggest this get separated out into a partial template. Have you seen how this is done, or would you need more support to do it? Thanks!

Comment (oorjitchowdhary) [reply app/views/layouts/application.html.erb]: Oh.. I'm so sorry @jywarren .. Didn't take a look here.. Will do the changes required now

**Comment by oorjitchowdhary** [comment]
I'm getting Missing Template error when I create a partial..
image

@publiclab/reviewers A little bit of help here..

**Comment by oorjitchowdhary** [comment]
@publiclab/reviewers Everything is done here.. These are the workings of the loader when not put in a partial.. I just need a little help with the creating of the partial and then we're good to go!

**Comment by oorjitchowdhary** [comment]
@SidharthBansal A little help here... I'm unable to create a partial template

**Comment by sashadev-sky** [comment]
@oorjitchowdhary try prepending an underscore to the partial name so that it looks like this: `app/views/layouts/_loading_page.html.erb`

**Comment by oorjitchowdhary** [comment]
Didn't work

**Comment by sashadev-sky** [comment]
@oorjitchowdhary note my above comments. Your code worked for me on my local repo after I made those small changes to the file names. I think you've got this!

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -80,10 +80,10 @@
   <link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/npm/leaflet-search@2.4.0/dist/leaflet-search.min.css">
   </head>
 
-  <body>
-
     <%= render :partial => 'layouts/header' %>
 
+   <%= render :partial => 'app/views/layouts/loading_page.html.erb' %>
```
Comment (sashadev-sky) [inline app/views/layouts/application.html.erb]: Per my last comment, try changing this to `<%= render :partial => 'layouts/loading_page' %> `

The path you use above is throwing an error for me.

File: app/views/layouts/_loading_page.html.erb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+  </body>
+
+<script>
+var myVar;
+ function myFunction() {
+    myVar = setTimeout(showPage, 3000);
+}
+ function showPage() {
+  document.getElementById("Loader").style.display = "none";
```
Comment (sashadev-sky) [inline app/views/layouts/_loading_page.html.erb]: - document.getElementById() is case sensitive so this is returning null
- try `document.getElementById("loader").style.display = "none"`;

**Comment by oorjitchowdhary** [comment]
Can anyone help me with Travis CI issues here...

File: app/views/layouts/application.html.erb [hunk 4]
```diff
@@ -80,10 +80,10 @@
   <link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/npm/leaflet-search@2.4.0/dist/leaflet-search.min.css">
   </head>
 
-  <body>
```
Comment (gauravano) [inline app/views/layouts/application.html.erb]: I don't think you are supposed to remove body tag. What do you think?

Comment (oorjitchowdhary) [reply app/views/layouts/application.html.erb]: I've added the body tag in the `loading_page` partial.. so I thought I should remove this one.. should I add it back?

Comment (gauravano) [reply app/views/layouts/application.html.erb]: Yes! for sure. It's the view of our main page. See, partials just get rendered at places where you put the render tag. You get better understanding from - https://guides.rubyonrails.org/layouts_and_rendering.html .

Comment (sashadev-sky) [reply app/views/layouts/application.html.erb]: It didn't change the results of my tests when I added the <body> back, but I saw that _loading_page.html.erb has another set of body tags inside it which I don't think should be there? I just don't know how to relate that back to the error messages

**Comment by oorjitchowdhary** [comment]
Let's see if this works

File: app/views/layouts/application.html.erb (hunk 4 above)
Comment (gauravano) [reply app/views/layouts/application.html.erb]: I also think here body tag might not be the reason as there is no if condition or so and DOM would eventually get complete with body tag in partial. I am checking if something else is wrong

Comment (oorjitchowdhary) [reply app/views/layouts/application.html.erb]: Right now through my latest commit, I've removed the body tags in the partial and added the body tags in the html code..

Comment (gauravano) [reply app/views/layouts/application.html.erb]: Looking into it!

**Comment by sashadev-sky** [comment]
@oorjitchowdhary I actually think my last suggestion might fix it! It got rid of the errors for me when I ran `rake test`. Sorry for the very long list of suggestions that is occurring here

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -81,9 +81,10 @@
   </head>
 
   <body>
-
     <%= render :partial => 'layouts/header' %>
```
[the remark this answers is not in the document]
Comment (gauravano) [reply app/views/layouts/application.html.erb]: Nice :mag: @sashadev-sky

**Comment by gauravano** [comment]
Also, you can keep the body tag in partial as DOM is completing and no test failure but I would not recommend it as we want our `application.html.erb` to be as clean and easy to understand as possible. Thanks!

And, yes test passed for the new commits.

**Comment by gauravano** [comment]
@sashadev-sky you can also save time by running a particular test file like `ruby -I test test/functional/wiki_controller_test.rb`

**Comment by oorjitchowdhary** [comment]
Yes, I won't add the `body` tags at the partial..

Checks passed... finally... 🎉 

Thank you everyone... @sashadev-sky , @gauravano and @SidharthBansal

**Comment by oorjitchowdhary** [comment]
So can this be merged now? Ⓜ️

**Comment by gauravano** [comment]
I will review this PR closely in the morning and then let you know further steps. Thanks!

**Comment by oorjitchowdhary** [comment]
@gauravano sure.. Gn 💤 for now guys.. Thanks once again

**Comment by sashadev-sky** [comment]
@gauravano got it. Sorry I removed the ready labels! 
@oorjitchowdhary thank you to both of you I learned from this one

**Comment by oorjitchowdhary** [comment]
@publiclab/reviewers Please take a look here. Thanks 😄

File: app/views/layouts/_loading_page.html.erb [hunk 6]
```diff
@@ -0,0 +1,32 @@
+<script>
+	onload="myFunction()" style="margin:0;">
```
Comment (jywarren) [inline app/views/layouts/_loading_page.html.erb]: Ah hmm, perhaps this has a copy-paste error? it seems to be supposed to be in an HTML tag? 

After all these changes, would one of you mind doing a few screencast gifs to demonstrate this working? Thanks a lot, and I appreciate your patience as we're extra careful with changes that will affect every page in the site. Thanks!!!

Comment (sashadev-sky) [reply app/views/layouts/_loading_page.html.erb]: @jywarren thank you for looking over the code 

@oorjitchowdhary would you like to make the updates and add the gifs?

**Comment by oorjitchowdhary** [comment]
Here are the gifs
loading 1

loading 2

loading 3

**Pull request closed without merging**